internal/lib/checkpoint.go

```diff
@@ -0,0 +1,233 @@
+		}
+	}
+	if err := c.storageRuntimeServer.StopContainer(ctr.ID()); err != nil {
+		return "", errors.Wrapf(err, "failed to unmount container %s", ctr.ID())
+	}
+	if err := c.ContainerStateToDisk(ctx, ctr); err != nil {
+		logrus.Warnf("Unable to write containers %s state to disk: %v", ctr.ID(), err)
+	}
+
+	if !opts.Keep {
+		cleanup := []string{
+			"dump.log",
+			"stats-dump",
+			metadata.ConfigDumpFile,
+			metadata.SpecDumpFile,
+		}
+		for _, del := range cleanup {
+			file := filepath.Join(ctr.Dir(), del)
+			if err := os.Remove(file); err != nil {
+				logrus.Debugf("Unable to remove file %s", file)
+			}
+		}
+	}
+
+	return ctr.ID(), nil
+}
+
+// Copied from libpod/diff.go
+var containerMounts = map[string]bool{
+	"/dev":               true,
+	"/etc/hostname":      true,
+	"/etc/hosts":         true,
+	"/etc/resolv.conf":   true,
+	"/proc":              true,
+	"/run":               true,
+	"/run/.containerenv": true,
+	"/run/secrets":       true,
+	"/sys":               true,
+}
+
+// getDiff returns the file system differences
+// Copied from libpod/diff.go and simplified for the checkpoint use case
+func (c *ContainerServer) getDiff(id string) ([]archive.Change, error) {
+	layerID, err := c.GetContainerTopLayerID(id)
+	if err != nil {
+		return nil, err
+	}
+	var rchanges []archive.Change
+	changes, err := c.store.Changes("", layerID)
+	if err == nil {
+		for _, c := range changes {
+			if containerMounts[c.Path] {
+				continue
+			}
+			rchanges = append(rchanges, c)
+		}
+	}
+	return rchanges, err
+}
+
+// To make the checkpoint/restore code use the same fields as Podman:
+type ContainerConfig struct {
+	ID              string `json:"id"`
+	Name            string `json:"name"`
+	RootfsImageName string `json:"rootfsImageName,omitempty"`
+}
+
+// prepareCheckpointExport writes the config and spec to
+// JSON files for later export
+// Podman: libpod/container_internal.go
+func (c *ContainerServer) prepareCheckpointExport(ctr *oci.Container) error {
+	config := &ContainerConfig{
+		ID:              ctr.ID(),
+		Name:            ctr.Name(),
+		RootfsImageName: ctr.ImageName(),
+	}
+
+	if _, err := metadata.WriteJSONFile(config, ctr.Dir(), metadata.ConfigDumpFile); err != nil {
+		return err
+	}
+
+	// save spec
+	jsonPath := filepath.Join(ctr.BundlePath(), "config.json")
+	g, err := generate.NewFromFile(jsonPath)
+	if err != nil {
+		return errors.Wrapf(err, "generating spec for container %q failed", ctr.ID())
+	}
+	if _, err := metadata.WriteJSONFile(g.Config, ctr.Dir(), metadata.SpecDumpFile); err != nil {
+		return errors.Wrapf(err, "generating spec for container %q failed", ctr.ID())
+	}
+
+	return nil
+}
+
+func (c *ContainerServer) exportCheckpoint(ctr *oci.Container, specgen *rspec.Spec, export string) error {
+	id := ctr.ID()
+	dest := ctr.Dir()
+	logrus.Debugf("Exporting checkpoint image of container %q to %q", id, dest)
+
+	includeFiles := []string{
+		"dump.log",
```

Comment:
I think this should be a `const` somewhere.

Reply:
done